### ISSUE TITLE
Audio

### ISSUE THREAD

**Issue by dk949** [tacticalhippopotamus/testing-app#4]
## 2 Types
* Sound effects
* Music (Find something copyright less)

**Comment by dk949** [comment]
## SoundEngine
* the `SoundEngine` should have a `play(Sample sample)` method, which will play a sound effect.
* when this is called a corresponding sample will play
    * they can be looked up in a `HashMap` (I think this works like `std::unordered_map` in c++, so it should not be too slow)
* music can be played in a much simpler way:
    * check if something is playing, if it is stop it
    * load the new track
    * play it
* volume
    * have methods `setSampleVolume(float volume)` and  `setTrackVolume(float volume)` 
* like textures, `Sound` and `Music` need to be disposed of. `dispose(Sample sample)` and `dispose(Track track)` methods may be necessary
    * although, most sounds will be used throughout the game, so is there a reason to dispose of them?

## Sample
* an enum listing all sound effects that can be played
* constants in the enum store the path to the file

## Track
* just like sample but for music

## When/how will the sounds be initialized?
* sound effects, preferably at launch, since there is enough latency in playing sounds as is
    * sounds can be initialized when `SoundEngine` is initialized
    * this would mean only one instance of `SoundEngine` should exists
* music, streamed from a file

## Rights to the music?
* yes this is a tiny project, but we should probably get this right anyway
* for sound effects, I have some that claim to be royalty free, but will need to look further into that 
* we can use Kevin MacLeod's music for the sound track
	* though we will need to credit him (which shouldn't be a problem?)
    * https://incompetech.com/music/royalty-free/faq.html 
    * Other options also exist

## Example of usage
```java
SoundEngine se = SoundEngine.getInstance();
se.play(Track.THEME);
se.play(Sample.WOOSH);

se.pauseTrack();
```

## Potential issues
* does this introduce global state?
    * yes, but everything is constant so it shouldn't be an issue
* should all sounds be accessible from anywhere in the program?
    * probably not, but I cannot figure out how to make that work without over complicating things
* should all sounds be loaded at once?
    * not sure

## File type?
### Initialization
|           | Run 1     | Run 2      | Run 3      | Run 4     | Average time | Standard deviation |
|-----------|-----------|------------|------------|-----------|--------------|---|
| mp3       | 9013530ns | 9660890ns  | 10375700ns | 4718340ns | 8.442115ms   | 2.544089805 |
| wav       | 7644350ns | 7388860ns  | 11154840ns | 7756720ns | 8.4861925ms  | 1.785744198 |
| ogg       | 7469040ns | 14569570ns | 8375840ns  | 6470960ns | 9.2213525ms  | 3.64936432 |
| Best time |           |            |            |           | 8.442115ms (mp3)  |   |




### Playing
|           | Run 1      | Run 2      | Run 3      | Run 4      | Average time | Standard deviation |
|-----------|------------|------------|------------|------------|--------------|---|
| mp3       | 19587830ns | 19014590ns | 24479550ns | 18545580ns | 20.4068875ms | 2.748356658 |
| wav       | 8352060ns  | 3630690ns  | 4067250ns  | 1731390ns  | 4.4453475ms  | 2.794911593 |
| ogg       | 15906010ns | 7127320ns  | 2181610ns  | 3592750ns  | 7.2019225ms  | 6.164315817 |
| Best time |            |            |            |            | 4.4453475ms  (wav) |   |

### In conclusion
* this is very unscientific
* also not tested on hardware (only emulator)
* best choice would probably be wav for sound effects and ogg/vorbis for music.
    * there used to be a licensing fee for using the mp3 format, but the patent seems to have expired.
    * ogg is allegedly better (also it's free), but everyone says there may be problems with it. switching from one to another is trivial.

## TODOs:
* there is no fade between tracks when tracks are switched.
* check platform specific sample rate, because it can vary. 
	* the emulator is already complaining about it
    * there is a way to change that dynamically, but for now everything is hard-coded (in the audio file) to 44.1kHz
